cmd/snap-update-ns/mount-entry.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "snap-update-ns/mount-entry.h"
+
+#include <errno.h>
+#include <mntent.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "../libsnap-confine-private/utils.h"
+#include "../libsnap-confine-private/string-utils.h"
+#include "../libsnap-confine-private/cleanup-funcs.h"
+
+/**
+ * Copy struct mntent into a freshly-allocated struct sc_mount_entry.
+ *
+ * The next pointer is initialized to NULL, it should be managed by the caller.
+ * If anything goes wrong the routine die()s.
+ **/
+static struct sc_mount_entry *sc_clone_mount_entry_from_mntent(const struct
+							       mntent *entry)
+{
+	struct sc_mount_entry *result;
+	result = calloc(1, sizeof *result);
+	if (result == NULL) {
+		die("cannot allocate memory");
+	}
+	result->mnt_fsname = strdup(entry->mnt_fsname);
+	if (result->mnt_fsname == NULL) {
+		die("cannot copy string");
+	}
+	result->mnt_dir = strdup(entry->mnt_dir);
+	if (result->mnt_dir == NULL) {
+		die("cannot copy string");
+	}
+	result->mnt_type = strdup(entry->mnt_type);
+	if (result->mnt_type == NULL) {
+		die("cannot copy string");
+	}
+	result->mnt_opts = strdup(entry->mnt_opts);
+	if (result->mnt_opts == NULL) {
+		die("cannot copy string");
+	}
+	result->mnt_freq = entry->mnt_freq;
+	result->mnt_passno = entry->mnt_passno;
+	return result;
+}
+
+static struct sc_mount_entry *sc_get_next_and_free_mount_entry(struct
+							       sc_mount_entry
+							       *entry)
+{
+	struct sc_mount_entry *next = entry->next;
+	free(entry->mnt_fsname);
+	free(entry->mnt_dir);
+	free(entry->mnt_type);
+	free(entry->mnt_opts);
+	memset(entry, 0, sizeof *entry);
+	free(entry);
+	return next;
+}
+
+void sc_free_mount_entry_list(struct sc_mount_entry *entry)
+{
+	while (entry != NULL) {
+		entry = sc_get_next_and_free_mount_entry(entry);
+	}
+}
+
+void sc_cleanup_mount_entry_list(struct sc_mount_entry **entryp)
+{
+	sc_free_mount_entry_list(*entryp);
```

Comment:
If a variable of any type has the cleanup attribute a function is called with a pointer to that variable. That is by definition not NULL. I've added the non-NULL checks there because someone may just use those functions directly. I guess that's a reason enough to add the check here but I wanted to make sure we agree on how the mechanism works.